[Chang Weng Yew Nicolas] iP

NicolasCwy opened the pull request:
# Kermit, your personal assistant

> "So, never worry about failure. It's GONNA happen. But that's okay: go ahead, take chances….and you might just find that > what feels like failure isn't failure at all, it's what inspires you to dream even bigger." 
> - Kermit The Frog

## Who is this for
* Suffering from crippling depression due to high school workload 😭 
* Hustlers who need a to keep track of their moola 💵 
* Procrastinators who want to feel good
No matter who you are, bask in Kermit's wisdom as he provides you wise words while keeping track of *YOUR* life

## Get Started
1. Download [ip.jar](https://github.com/NicolasCwy/ip/releases/tag/A-Jar)
2. Put it into a folder that you want to be the root directory for Kermit!
3. Double click the jar
4. 💵 💵 Get your life in check 💵 💵 

## Features
- [x] Saving tasks to computer
- [x] Some error handling - Kermit isn't all knowing😳
- [x] Supports todo tasks, deadlines and events!
- [x] Filter support for tasks based on keywords!
- [x] Date support
- [ ] Time support

## Enter the Matrix
Run the main function in the following file if you cloned the repo!
```java
    // Kermit.java
    public static void main(String[] args){
        new Kermit("data/tasks.txt").run();
    }
```
If it works correctly Kermit should say `Hello I am Kermit ( *・∀・)ノ゛, eaten any flies today?
What can I do for you?`

Commit:
Add Gradle support

NicolasCwy committed:
Rename Duke file to Kermit

NicolasCwy committed:
Add helper function to format text

NicolasCwy committed:
Add echo function to Kermit Bot

NicolasCwy committed:
Fix grammar error in intro text

NicolasCwy committed:
Change echo to add text

NicolasCwy committed:
Change access modifier of formatText() to private

formatText is used to format Kermit's speech and should not be used outside the class

Change formatText() method access modifier to private

formatText being private hides the implementation details from users

NicolasCwy committed:
Add list command to Kermit

Typing to text (with the exception of "bye" and "list") would add the text to a list. The list command would print all the items is in the list

NicolasCwy committed:
Move list logic to ToDo class

NicolasCwy committed:
Add class to encapsulate task

NicolasCwy committed:
Rename variables to be less ambiguous

NicolasCwy committed:
Make ToDo list use Task class

NicolasCwy committed:
Remove generic from ToDo

Remove type parameter from ToDo class

NicolasCwy committed:
Add completeTask method

Marks task in ToDo list as complete

NicolasCwy committed:
Add done command to Kermit bot

Mark items in todo list as complete using "done" command. This is reflected in the "list" method as well

NicolasCwy committed:
Add speech when Kermit prints list

NicolasCwy committed:
Change logic to use StringBuilder class

StringBuilder object used to construct strings when reading the input commands issued by user

NicolasCwy committed:
Add todos task to Kermit

todos encapsulates simple tasks that only have a description and can be completed.
To prepare for more complex tasks such as deadline and event

NicolasCwy committed:
Change Kermit control flow to use switch statement

Return is used in "bye" case as break would exit out of the switch structure. Return is needed to exit the while loop and stop the program

NicolasCwy committed:
Line wrapped speech statement

NicolasCwy committed:
Add deadline task

Deadline is a task that has a deadline

NicolasCwy committed:
Add deadline command to Kermit bot

NicolasCwy committed:
Add Event task class

NicolasCwy committed:
Change deadline string

Deadline toString() returns a prefix "[D]" instead of an "[E]" which represented an event task

NicolasCwy committed:
Add helper function to create message when task is added

Helper function creates message to be used

String message was too long and was repeated 3 times in the code. Helper function used to reduce code re-use

NicolasCwy committed:
Amend missing spaces in Kermit speech

NicolasCwy committed:
Add expected and input test for semi-automated test

NicolasCwy committed:
Change script file reference class

Script redirects input to bin/Kermit instead of bin/Duke

Script was referring to the original name of the iP bot Duke. Running this version would cause no output to be redirected to the file ACTUAL.txt for testing, causing the test to fail.

NicolasCwy committed:
Change invalid command text

NicolasCwy committed:
Add KermitException for Kermit bot errors

KermitException should be used for any errors that occur during the runtime of Kermit Bot

NicolasCwy committed:
Throw KermitException for invalid command

Invalid command text throws KermitException

An invalid command should raise an exception instead of being handled in the switch statement as an expected behaviour and just being printed out.

NicolasCwy committed:
Add error checking to Kermit

Handle empty description for events and invalid flag arguments

NicolasCwy committed:
Add test cases to test exceptions

NicolasCwy committed:
Make error thrown for missing flag argument Event clearer

NicolasCwy committed:
Extend horizontal divider for styling

NicolasCwy committed:
Add string at start of every error

Add string that will be printed before every error message, to add flavour to every error message. Make it consistent so that it is recognisable,

NicolasCwy committed:
Check validity of command outside switch statement

NicolasCwy committed:
Fix expected file

NicolasCwy committed:
Move error checking logic out of switch statement

Error checking is done as description and flag are constructed to reduce code repetition.

Ensures that inputs are valid before executing command logic

NicolasCwy committed:
Add test case for delete command

NicolasCwy committed:
removeTask: return task

NicolasCwy committed:
Add delete command (no error checking)

NicolasCwy committed:
Add creation of Kermit data file

NicolasCwy committed:
Add methods for file writing format

NicolasCwy committed:
Create subclass for date dependent tasks

NicolasCwy committed:
Add helper function to format save strings

NicolasCwy committed:
Add printwriter

NicolasCwy committed:
Fix formatWriteString not changing string

NicolasCwy committed:
Add save user data to file

NicolasCwy committed:
Remove catch for RuntimeException

NicolasCwy committed:
Rename variable to be less ambiguous

NicolasCwy committed:
Add constructor for tasks that are completed

NicolasCwy committed:
Add read from existing todo list data file capability

NicolasCwy committed:
Change test for new localdate format

NicolasCwy committed:
Change to accept localdate time

NicolasCwy committed:
Change kermit to use localdate for dates

NicolasCwy committed:
Add helper function to format and print text

NicolasCwy committed:
Merge branch 'branch-Level-7'

NicolasCwy committed:
Merge branch 'branch-Level-8'

NicolasCwy committed:
Fix incorrect datetime format

NicolasCwy committed:
Change common string output to task class

NicolasCwy committed:
Add test cases for find command

(cherry picked from commit fa4fde4b08ce01011154a30cdd0f8158702bb4f3)

NicolasCwy committed:
Add filter method to ToDo list

(cherry picked from commit ecd9e53164c1f754c38943c72a3a64a57c8361c3)

NicolasCwy committed:
Add filter task output string

(cherry picked from commit 931d60c4b2a04073e94d13a4d843b281d143b481)

NicolasCwy committed:
Fix incorrect return type

(cherry picked from commit 460246e8e2cc8e89d33f83beef51dfcfb849636a)

NicolasCwy committed:
Fix expected test case

(cherry picked from commit 706b1ea99e56b5315a5d13c8742543a0d70223b9)

NicolasCwy committed:
Add find command to Kermit

(cherry picked from commit 5a98bdc9cf1cb4acb77535a5302386404ff04987)

NicolasCwy committed:
Change code to use a storage class in an ideal scenario

NicolasCwy committed:
Add storage class

NicolasCwy committed:
Move read from file to Storage class

NicolasCwy committed:
Expand try block in Kermit

NicolasCwy committed:
Add save function to storage

NicolasCwy committed:
Add Ui Class

NicolasCwy committed:
Fix unable to read from existing data file bug

NicolasCwy committed:
Make Kermit use Ui class

NicolasCwy committed:
Add loadingError text option to Ui

NicolasCwy committed:
Add constructor for ToDo that accepts lists

NicolasCwy committed:
Change Storage to return an array of tasks after load

NicolasCwy committed:
Compartmentalise code to use instance of Kermit

NicolasCwy committed:
Remove comment in Storage class

NicolasCwy committed:
Change file loading error text

NicolasCwy committed:
Catch exception if data file does not follow format

NicolasCwy committed:
Move scan user input into Ui

NicolasCwy committed:
Add Command class

NicolasCwy committed:
Add execute abstract method

NicolasCwy committed:
Change execute parameters

NicolasCwy committed:
Add KermitException throw to execute method signature

NicolasCwy committed:
Add Exit command

NicolasCwy committed:
Add ListTask command

NicolasCwy committed:
Remove thrown exception from overridden execute signature

NicolasCwy committed:
Add complete task command

NicolasCwy committed:
Add delete task command

NicolasCwy committed:
Add "Add todo" command

NicolasCwy committed:
Add "add date dependent task" command

NicolasCwy committed:
Change Ui argument to be string

NicolasCwy committed:
Change structure Kermit to be more OOP

NicolasCwy committed:
Change constructor to take in one-indexed task numbers and change error

NicolasCwy committed:
Change expected error message

NicolasCwy committed:
Add parser class

NicolasCwy committed:
Internalise argument checking into commands

NicolasCwy committed:
Fix date parsing

NicolasCwy committed:
Change expected output

NicolasCwy committed:
Rename command classes to have command in name

NicolasCwy committed:
Package command files

NicolasCwy committed:
Package task files

NicolasCwy committed:
Package Kermit related files

NicolasCwy committed:
Fix incorrect access modifier

NicolasCwy committed:
Fix IDE renaming error

NicolasCwy committed:
Fix IDE renaming error

NicolasCwy committed:
Change testing file to run on packaged files

NicolasCwy committed:
Add ToDo Test

NicolasCwy committed:
Add getShortForm test

NicolasCwy committed:
Add Event test file

NicolasCwy committed:
Fix GetShortForm test

NicolasCwy committed:
Add event toString test

NicolasCwy committed:
Add deadline test

NicolasCwy committed:
Add javadocs for Kermit

NicolasCwy committed:
Add javadocs for KermitException

NicolasCwy committed:
Add javadocs for Parser

NicolasCwy committed:
Remove redundant command check

NicolasCwy committed:
Remove wildcard import and import classes

NicolasCwy committed:
Add javadocs to storage class

NicolasCwy committed:
Add javadocs to ToDo class

NicolasCwy committed:
Add periods at end of javadoc headers.

NicolasCwy committed:
Add javadocs to Ui

NicolasCwy committed:
Add javadocs to DateDependentTask

NicolasCwy committed:
Add period at end of javadoc header

NicolasCwy committed:
Add javadocs to deadline

NicolasCwy committed:
Remove redundant getDateString

getDateString is deleted as it is inherited from DateDependantTask,

NicolasCwy committed:
Add javadocs to event

NicolasCwy committed:
Add override tag to methods

NicolasCwy committed:
Add javadocs to ToDo

NicolasCwy committed:
Add override tag to method

NicolasCwy committed:
Add periods at end of javadoc header

NicolasCwy committed:
Add javadocs to Task

NicolasCwy committed:
Add javadocs for Command

NicolasCwy committed:
Add javadocs for exit command

NicolasCwy committed:
Add javadocs for complete task command

NicolasCwy committed:
Add override annotation

NicolasCwy committed:
Remove comment

NicolasCwy committed:
Make javadoc less ambiguous

NicolasCwy committed:
Add javadocs for delete task command.

NicolasCwy committed:
Add javadocs for list tasks command.

NicolasCwy committed:
Add javadocs for add tasks command.

NicolasCwy committed:
Properly indent switch cases

NicolasCwy committed:
Add proper access modifier

NicolasCwy committed:
Add proper access modifier

NicolasCwy committed:
Add missing override annotation

NicolasCwy committed:
Add proper access modifer

NicolasCwy committed:
Rename getDate and getDatestring methods

Renamed getDate method to getDateString. Renamed getDateString to getFormattedDateString

NicolasCwy committed:
Add override annotation

NicolasCwy committed:
Add override annotation

NicolasCwy committed:
Remove unnecessary method

Removed getFormattedDateString() in Deadline as it calls the parent function without any changes

NicolasCwy committed:
Rename ToDo to TaskList

NicolasCwy committed:
Rename completeTask to markTaskAsComplete

NicolasCwy committed:
Delete formatUserDateFormat
function

formatUserDateFormat is deleted from Kermit.

formatUserDateFormat is superseded by Parser class and does not handle date parsing by user.

NicolasCwy committed:
Indent switch case

NicolasCwy committed:
Change variable assigned from flag to dateString

Assigned value to dateString instead of flag.

flag is meant to be the keyword after the description e.g /by or /at and the dateString is the date that follows.

NicolasCwy committed:
Change method name completeTask to markTaskAsComplete

Method name was changed in all files a few commits ago, but the method name in TaskList did not change

NicolasCwy committed:
Indent switch case

NicolasCwy committed:
Add missing constructor method

NicolasCwy committed:
Reduce line length to less than 120

NicolasCwy committed:
Rename ToDos to ToDo

NicolasCwy committed:
Merge branch 'branch-A-CodingStandard'

NicolasCwy committed:
Merge branch 'branch-A-JavaDoc'

NicolasCwy committed:
Add missing javadocs

NicolasCwy committed:
Merge branch 'branch-Level-9'

NicolasCwy committed:
Merge branch 'add-gradle-support' into A-Gradle

NicolasCwy committed:
Change main class name for gradle to build project

NicolasCwy committed:
Change base archive name, to use shadowJar

NicolasCwy committed:
Add checkstyle.xml from adb3

NicolasCwy committed:
Fix checkstyle violations

NicolasCwy committed:
Fix checkstyle violations

lwlshawn reviewed (commented):
Your code looks great! I didn't manage to spot any coding standard issues that slipped past in the files i looked at.

tsm1820 reviewed (commented):
LGTM. Good Job!! The variable and method names used all appear to follow the naming conventions and the coding standard rules are also well-observed :)

NicolasCwy commented:
@lwlshawn @tsm1820 Thanks for helping to review guys!

NicolasCwy committed:
Add JavaFX as a dependency

NicolasCwy committed:
Add pictures for GUI

NicolasCwy committed:
Add initial GUI from SeEdu tutorial

NicolasCwy committed:
Convert GUI to use FXML

NicolasCwy committed:
Fix font and replace it with Times New Roman

NicolasCwy committed:
Change JavaFX version from 16 to 11

NicolasCwy committed:
Move getResponse into kermit class

NicolasCwy committed:
Amend code structure to be used in the GUI

NicolasCwy committed:
Amend javadocs for UI class and delete unnecessary methods

NicolasCwy committed:
Add new line according to checkstyle

NicolasCwy committed:
Delete un-used run method

NicolasCwy committed:
Add title to window

NicolasCwy committed:
Add intro message

NicolasCwy committed:
Close window after bye command

hungkhoaitay reviewed (commented):
The code is well written, I couldn't find and coding standard violations. Great job! 💯

lwlshawn commented on src/main/java/kermit/Ui.java:
```diff
@@ -0,0 +1,113 @@
+    public Ui() {
+    }
+
+    /**
```
Sorry for commenting on this again! The course required leaving a review comment instead of a normal comment. Still think your code is great but also thought it was impressive how much documentation you provided.

lwlshawn reviewed (commented):
For the 2103 script

NicolasCwy committed:
Add intro message that was missing

NicolasCwy committed:
Point gradle to new program insertion point

NicolasCwy committed:
Add assertion

NicolasCwy committed:
Delete redundant flag variable

Flag variable is not used throughout the program, no checking is done.

Flag variable is deleted to prevent confusion

Flag variable points to the name of the flag e.g /by or /at
from the full command given by the user.

NicolasCwy committed:
Add basic CI to project

NicolasCwy committed:
Refactor parser

Code duplication was found when constructing the
flag and the task description from a users input. The parse method which also did the parsing was long.

Divide code into smaller methods to reduce method length. Remove the use of StringBuilder and used Arrays class.

Shorter methods and a KISS approach(by replacing StringBuilder) allow for easier debugging in the future.

NicolasCwy committed:
Refactor code

NicolasCwy committed:
Merge pull request #2 from NicolasCwy/branch-A-Assertions

Add assertion to code

NicolasCwy committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/kermit/Parser.java

NicolasCwy committed:
Merge pull request #1 from NicolasCwy/branch-A-CodeQuality

Refactored files based on CS2103T code quality section

NicolasCwy committed:
Add help command

NicolasCwy committed:
Add help syntax for most commands

NicolasCwy committed:
Add help syntax for all commands

NicolasCwy committed:
Change task filter implementation

NicolasCwy committed:
Fix cut dialog boxes

NicolasCwy committed:
Add colour for user and kermit text-box

NicolasCwy committed:
Add border radius and padding to dialogbox

NicolasCwy committed:
Set theme jekyll-theme-cayman

NicolasCwy committed:
Fix deadline bug

NicolasCwy committed:
Add product ui picture

NicolasCwy committed:
Merge remote-tracking branch 'origin/master'

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Change Kermit and User profile

NicolasCwy committed:
Change GitHub page photo

NicolasCwy committed:
Update README.md

Change kermit get-started picture

NicolasCwy committed:
Merge remote-tracking branch 'origin/master'

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Merge remote-tracking branch 'origin/master'

NicolasCwy committed:
Add command usage

NicolasCwy committed:
Merge branch 'branch-A-UserGuide'

NicolasCwy committed:
Change readme

NicolasCwy committed:
Change readme

NicolasCwy committed:
Merge branch 'branch-A-UserGuide'

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Update README.md

NicolasCwy committed:
Change user profile pic

NicolasCwy committed:
Add javadocs for all classes

NicolasCwy committed:
Remove redundant CLI test

NicolasCwy committed:
Remove CLI test

NicolasCwy committed:
Apply SLAP to storage class

NicolasCwy committed:
Change import order
